### PR TITLE
AI Chat: Tweaked article length truncation

### PR DESCRIPTION
**Pull request by nullhook**
Resolves https://github.com/brave/brave-browser/issues/32346

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Review by petemill — CHANGES_REQUESTED** [review]
As discussed, I don't think this is the correct place to do the truncating since it won't include video transcript urls or post-processing (e.g youtube xml parsing). I think it could be good to do this inside `AIChatTabHelper::OnTabContentRetrieved` since that's where all content is received. It also has access to the model being used, so it could use dynamic limits (not necessarily in this PR).

File: components/ai_chat/renderer/page_text_distilling.cc [hunk 1]
```diff
@@ -132,11 +132,9 @@ void DistillPageText(
...
-  // 8092 chars, ~3,098 tokens (reserved for article)
-  // 1k chars, ~380 tokens (reserved for prompt)
-  std::string contents_text = base::UTF16ToUTF8(
-      base::JoinString(text_node_contents, u" ").substr(0, 8092));
+  std::string contents_text =
+      base::UTF16ToUTF8(base::JoinString(text_node_contents, u" "));
+
   if (contents_text.empty()) {
     std::move(callback).Run({});
     return;
```
Comment (nullhook) [inline components/ai_chat/renderer/page_text_distilling.cc:140]: We do a check if text is empty in `OnTabContentRetrieved` do we need another one here?

File: components/ai_chat/browser/ai_chat_tab_helper.cc [hunk 2]
```diff
@@ -196,6 +196,12 @@ void AIChatTabHelper::OnTabContentRetrieved(int64_t for_navigation_id,
     VLOG(1) << __func__ << ": No data";
     return;
   }
+
+  // tokens + max_new_tokens must be <= 4096 (llama2)
+  // 8092 chars, ~3,098 tokens (reserved for article)
+  // 1k chars, ~380 tokens (reserved for prompt)
+  contents_text = contents_text.substr(0, 8092);
+
```
Comment (petemill) [inline components/ai_chat/browser/ai_chat_tab_helper.cc:204]: (and here we can set a bool flag to indicate to the user that only a part of the page is being shared with leo as it's too long, in the subsequent PR)

File: components/ai_chat/renderer/page_text_distilling.cc (hunk 1 above)
Comment (petemill) [reply components/ai_chat/renderer/page_text_distilling.cc:140]: no

Comment (petemill) [reply components/ai_chat/renderer/page_text_distilling.cc:140]: maybe a limit on the parser so it doesn't get wild, but I think we already have that with the Snapshotter, which limits to 9000 nodes and 4 seconds. And we also have a wildness limit on the video transcript fetching in terms of Mb fetched.

**Review by petemill — APPROVED** [review]

**Pull request merged**